src/System.Private.ServiceModel/src/System/ServiceModel/Channels/RequestChannel.cs

```diff
@@ -196,7 +196,8 @@ private void ReleaseRequest(IRequestBase request)
                 _outstandingRequests.Remove(request);
                 if (_outstandingRequests.Count == 0)
                 {
-                    if (!_closed && _closedTcs != null)
+                    // When we are closed or closing, _closedTcs is managed by the close logic.
+                    if (!_closed && State != CommunicationState.Closing && _closedTcs != null)
```

Comment:
You still have a race condition here, it's just a lot tighter. _closedTcs could be not null in the if expression, and then null below. I would replace the body with:
```csharp
var closedTcs = Interlocked.CompareExchange(ref _closedTcs, null, _closedTcs);
if (closedTcs != null)
{
    closedTcs.TrySetResult(null);
}
```